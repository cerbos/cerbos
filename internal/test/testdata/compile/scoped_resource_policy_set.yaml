---
mainDef: "leave_request.acme.hr.uk.yaml"
inputDefs:
  "leave_request.yaml":
     apiVersion: api.cerbos.dev/v1
     resourcePolicy:
       resource: leave_request
       version: "20210210"
       importDerivedRoles:
         - my_derived_roles
       schemas:
         principalSchema:
           ref: cerbos:///complex_object.json
         resourceSchema:
           ref: cerbos:///complex_object.json
       rules:
         - actions: ['*']
           effect: EFFECT_ALLOW
           roles:
             - admin
           name: wildcard

  "leave_request.acme.yaml":
     apiVersion: api.cerbos.dev/v1
     resourcePolicy:
       resource: leave_request
       version: "20210210"
       scope: "acme"
       importDerivedRoles:
         - my_derived_roles
       schemas:
         principalSchema:
           ref: cerbos:///complex_object.json
         resourceSchema:
           ref: cerbos:///complex_object.json
       rules:
         - actions: ["create"]
           derivedRoles:
             - employee_that_owns_the_record
           effect: EFFECT_ALLOW
         - actions: ["view:*"]
           derivedRoles:
             - employee_that_owns_the_record
             - direct_manager
           effect: EFFECT_ALLOW
         - actions: ["view:public"]
           derivedRoles:
             - any_employee
           effect: EFFECT_ALLOW
         - actions: ["approve"]
           condition:
             match:
               expr: request.resource.attr.status == "PENDING_APPROVAL"
           derivedRoles:
             - direct_manager
           effect: EFFECT_ALLOW

  "leave_request.acme.hr.yaml":
     apiVersion: api.cerbos.dev/v1
     resourcePolicy:
       resource: leave_request
       version: "20210210"
       scope: "acme.hr"
       importDerivedRoles:
         - my_derived_roles
       rules:
         - actions: ['*']
           effect: EFFECT_ALLOW
           roles:
             - admin
             - hr_admin
           name: wildcard

  "leave_request.acme.hr.uk.yaml":
     apiVersion: api.cerbos.dev/v1
     resourcePolicy:
       resource: leave_request
       version: "20210210"
       scope: "acme.hr.uk"
       importDerivedRoles:
         - my_derived_roles
       rules:
         - actions: ["approve"]
           condition:
             match:
               expr: request.resource.attr.status == "PENDING_APPROVAL"
           derivedRoles:
             - direct_manager
           roles:
             - hr_bp
           effect: EFFECT_ALLOW

  "my_derived_roles.yaml":
     apiVersion: "api.cerbos.dev/v1"
     derivedRoles:
       name: my_derived_roles
       definitions:
         - name: tester
           parentRoles: ["dev", "qa"]

         - name: employee_that_owns_the_record
           parentRoles: ["employee"]
           condition:
             match:
               expr: R.attr.owner == P.id

         - name: any_employee
           parentRoles: ["employee"]

         - name: direct_manager
           parentRoles: ["manager"]
           condition:
             match:
               all:
                 of:
                   - expr: "request.resource.attr.geography == request.principal.attr.geography"
                   - expr: "request.resource.attr.geography == request.principal.attr.managed_geographies"
