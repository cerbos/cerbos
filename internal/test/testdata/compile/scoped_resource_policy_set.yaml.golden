{
  "fqn": "cerbos.resource.leave_request.v20210210/acme.hr.uk",
  "resourcePolicy": {
    "meta": {
      "fqn": "cerbos.resource.leave_request.v20210210/acme.hr.uk",
      "resource": "leave_request",
      "version": "20210210",
      "sourceAttributes": {
        "resource.leave_request.v20210210": {
          "attributes": {
            "source": "leave_request.yaml"
          }
        },
        "resource.leave_request.v20210210/acme": {
          "attributes": {
            "source": "leave_request.acme.yaml"
          }
        },
        "resource.leave_request.v20210210/acme.hr": {
          "attributes": {
            "source": "leave_request.acme.hr.yaml"
          }
        },
        "resource.leave_request.v20210210/acme.hr.uk": {
          "attributes": {
            "source": "leave_request.acme.hr.uk.yaml"
          }
        }
      }
    },
    "policies": [
      {
        "scope": "acme.hr.uk",
        "derivedRoles": {
          "direct_manager": {
            "name": "direct_manager",
            "parentRoles": {
              "manager": {}
            },
            "condition": {
              "all": {
                "expr": [
                  {
                    "expr": {
                      "original": "request.resource.attr.geography == request.principal.attr.geography",
                      "checked": {
                        "referenceMap": {
                          "1": {
                            "name": "request"
                          },
                          "5": {
                            "overloadId": [
                              "equals"
                            ]
                          },
                          "6": {
                            "name": "request"
                          }
                        },
                        "typeMap": {
                          "1": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "2": {
                            "messageType": "cerbos.engine.v1.Request.Resource"
                          },
                          "3": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "4": {
                            "dyn": {}
                          },
                          "5": {
                            "primitive": "BOOL"
                          },
                          "6": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "7": {
                            "messageType": "cerbos.engine.v1.Request.Principal"
                          },
                          "8": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "9": {
                            "dyn": {}
                          }
                        },
                        "sourceInfo": {
                          "location": "<input>",
                          "lineOffsets": [
                            68
                          ],
                          "positions": {
                            "1": 0,
                            "2": 7,
                            "3": 16,
                            "4": 21,
                            "5": 32,
                            "6": 35,
                            "7": 42,
                            "8": 52,
                            "9": 57
                          }
                        },
                        "expr": {
                          "id": "5",
                          "callExpr": {
                            "function": "_==_",
                            "args": [
                              {
                                "id": "4",
                                "selectExpr": {
                                  "operand": {
                                    "id": "3",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "2",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "1",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "resource"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              },
                              {
                                "id": "9",
                                "selectExpr": {
                                  "operand": {
                                    "id": "8",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "7",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "6",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "principal"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  },
                  {
                    "expr": {
                      "original": "request.resource.attr.geography == request.principal.attr.managed_geographies",
                      "checked": {
                        "referenceMap": {
                          "1": {
                            "name": "request"
                          },
                          "5": {
                            "overloadId": [
                              "equals"
                            ]
                          },
                          "6": {
                            "name": "request"
                          }
                        },
                        "typeMap": {
                          "1": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "2": {
                            "messageType": "cerbos.engine.v1.Request.Resource"
                          },
                          "3": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "4": {
                            "dyn": {}
                          },
                          "5": {
                            "primitive": "BOOL"
                          },
                          "6": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "7": {
                            "messageType": "cerbos.engine.v1.Request.Principal"
                          },
                          "8": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "9": {
                            "dyn": {}
                          }
                        },
                        "sourceInfo": {
                          "location": "<input>",
                          "lineOffsets": [
                            78
                          ],
                          "positions": {
                            "1": 0,
                            "2": 7,
                            "3": 16,
                            "4": 21,
                            "5": 32,
                            "6": 35,
                            "7": 42,
                            "8": 52,
                            "9": 57
                          }
                        },
                        "expr": {
                          "id": "5",
                          "callExpr": {
                            "function": "_==_",
                            "args": [
                              {
                                "id": "4",
                                "selectExpr": {
                                  "operand": {
                                    "id": "3",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "2",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "1",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "resource"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              },
                              {
                                "id": "9",
                                "selectExpr": {
                                  "operand": {
                                    "id": "8",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "7",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "6",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "principal"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "managed_geographies"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  }
                ]
              }
            }
          }
        },
        "rules": [
          {
            "name": "rule-001",
            "actions": {
              "approve": {}
            },
            "derivedRoles": {
              "direct_manager": {}
            },
            "roles": {
              "hr_bp": {}
            },
            "condition": {
              "expr": {
                "original": "request.resource.attr.status == \"PENDING_APPROVAL\"",
                "checked": {
                  "referenceMap": {
                    "1": {
                      "name": "request"
                    },
                    "5": {
                      "overloadId": [
                        "equals"
                      ]
                    }
                  },
                  "typeMap": {
                    "1": {
                      "messageType": "cerbos.engine.v1.Request"
                    },
                    "2": {
                      "messageType": "cerbos.engine.v1.Request.Resource"
                    },
                    "3": {
                      "mapType": {
                        "keyType": {
                          "primitive": "STRING"
                        },
                        "valueType": {
                          "dyn": {}
                        }
                      }
                    },
                    "4": {
                      "dyn": {}
                    },
                    "5": {
                      "primitive": "BOOL"
                    },
                    "6": {
                      "primitive": "STRING"
                    }
                  },
                  "sourceInfo": {
                    "location": "<input>",
                    "lineOffsets": [
                      51
                    ],
                    "positions": {
                      "1": 0,
                      "2": 7,
                      "3": 16,
                      "4": 21,
                      "5": 29,
                      "6": 32
                    }
                  },
                  "expr": {
                    "id": "5",
                    "callExpr": {
                      "function": "_==_",
                      "args": [
                        {
                          "id": "4",
                          "selectExpr": {
                            "operand": {
                              "id": "3",
                              "selectExpr": {
                                "operand": {
                                  "id": "2",
                                  "selectExpr": {
                                    "operand": {
                                      "id": "1",
                                      "identExpr": {
                                        "name": "request"
                                      }
                                    },
                                    "field": "resource"
                                  }
                                },
                                "field": "attr"
                              }
                            },
                            "field": "status"
                          }
                        },
                        {
                          "id": "6",
                          "constExpr": {
                            "stringValue": "PENDING_APPROVAL"
                          }
                        }
                      ]
                    }
                  }
                }
              }
            },
            "effect": "EFFECT_ALLOW"
          }
        ]
      },
      {
        "scope": "acme.hr",
        "rules": [
          {
            "name": "wildcard",
            "actions": {
              "*": {}
            },
            "roles": {
              "admin": {},
              "hr_admin": {}
            },
            "effect": "EFFECT_ALLOW"
          }
        ]
      },
      {
        "scope": "acme",
        "derivedRoles": {
          "any_employee": {
            "name": "any_employee",
            "parentRoles": {
              "employee": {}
            }
          },
          "direct_manager": {
            "name": "direct_manager",
            "parentRoles": {
              "manager": {}
            },
            "condition": {
              "all": {
                "expr": [
                  {
                    "expr": {
                      "original": "request.resource.attr.geography == request.principal.attr.geography",
                      "checked": {
                        "referenceMap": {
                          "1": {
                            "name": "request"
                          },
                          "5": {
                            "overloadId": [
                              "equals"
                            ]
                          },
                          "6": {
                            "name": "request"
                          }
                        },
                        "typeMap": {
                          "1": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "2": {
                            "messageType": "cerbos.engine.v1.Request.Resource"
                          },
                          "3": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "4": {
                            "dyn": {}
                          },
                          "5": {
                            "primitive": "BOOL"
                          },
                          "6": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "7": {
                            "messageType": "cerbos.engine.v1.Request.Principal"
                          },
                          "8": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "9": {
                            "dyn": {}
                          }
                        },
                        "sourceInfo": {
                          "location": "<input>",
                          "lineOffsets": [
                            68
                          ],
                          "positions": {
                            "1": 0,
                            "2": 7,
                            "3": 16,
                            "4": 21,
                            "5": 32,
                            "6": 35,
                            "7": 42,
                            "8": 52,
                            "9": 57
                          }
                        },
                        "expr": {
                          "id": "5",
                          "callExpr": {
                            "function": "_==_",
                            "args": [
                              {
                                "id": "4",
                                "selectExpr": {
                                  "operand": {
                                    "id": "3",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "2",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "1",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "resource"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              },
                              {
                                "id": "9",
                                "selectExpr": {
                                  "operand": {
                                    "id": "8",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "7",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "6",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "principal"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  },
                  {
                    "expr": {
                      "original": "request.resource.attr.geography == request.principal.attr.managed_geographies",
                      "checked": {
                        "referenceMap": {
                          "1": {
                            "name": "request"
                          },
                          "5": {
                            "overloadId": [
                              "equals"
                            ]
                          },
                          "6": {
                            "name": "request"
                          }
                        },
                        "typeMap": {
                          "1": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "2": {
                            "messageType": "cerbos.engine.v1.Request.Resource"
                          },
                          "3": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "4": {
                            "dyn": {}
                          },
                          "5": {
                            "primitive": "BOOL"
                          },
                          "6": {
                            "messageType": "cerbos.engine.v1.Request"
                          },
                          "7": {
                            "messageType": "cerbos.engine.v1.Request.Principal"
                          },
                          "8": {
                            "mapType": {
                              "keyType": {
                                "primitive": "STRING"
                              },
                              "valueType": {
                                "dyn": {}
                              }
                            }
                          },
                          "9": {
                            "dyn": {}
                          }
                        },
                        "sourceInfo": {
                          "location": "<input>",
                          "lineOffsets": [
                            78
                          ],
                          "positions": {
                            "1": 0,
                            "2": 7,
                            "3": 16,
                            "4": 21,
                            "5": 32,
                            "6": 35,
                            "7": 42,
                            "8": 52,
                            "9": 57
                          }
                        },
                        "expr": {
                          "id": "5",
                          "callExpr": {
                            "function": "_==_",
                            "args": [
                              {
                                "id": "4",
                                "selectExpr": {
                                  "operand": {
                                    "id": "3",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "2",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "1",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "resource"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "geography"
                                }
                              },
                              {
                                "id": "9",
                                "selectExpr": {
                                  "operand": {
                                    "id": "8",
                                    "selectExpr": {
                                      "operand": {
                                        "id": "7",
                                        "selectExpr": {
                                          "operand": {
                                            "id": "6",
                                            "identExpr": {
                                              "name": "request"
                                            }
                                          },
                                          "field": "principal"
                                        }
                                      },
                                      "field": "attr"
                                    }
                                  },
                                  "field": "managed_geographies"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  }
                ]
              }
            }
          },
          "employee_that_owns_the_record": {
            "name": "employee_that_owns_the_record",
            "parentRoles": {
              "employee": {}
            },
            "condition": {
              "expr": {
                "original": "R.attr.owner == P.id",
                "checked": {
                  "referenceMap": {
                    "1": {
                      "name": "R"
                    },
                    "4": {
                      "overloadId": [
                        "equals"
                      ]
                    },
                    "5": {
                      "name": "P"
                    }
                  },
                  "typeMap": {
                    "1": {
                      "messageType": "cerbos.engine.v1.Request.Resource"
                    },
                    "2": {
                      "mapType": {
                        "keyType": {
                          "primitive": "STRING"
                        },
                        "valueType": {
                          "dyn": {}
                        }
                      }
                    },
                    "3": {
                      "dyn": {}
                    },
                    "4": {
                      "primitive": "BOOL"
                    },
                    "5": {
                      "messageType": "cerbos.engine.v1.Request.Principal"
                    },
                    "6": {
                      "primitive": "STRING"
                    }
                  },
                  "sourceInfo": {
                    "location": "<input>",
                    "lineOffsets": [
                      21
                    ],
                    "positions": {
                      "1": 0,
                      "2": 1,
                      "3": 6,
                      "4": 13,
                      "5": 16,
                      "6": 17
                    }
                  },
                  "expr": {
                    "id": "4",
                    "callExpr": {
                      "function": "_==_",
                      "args": [
                        {
                          "id": "3",
                          "selectExpr": {
                            "operand": {
                              "id": "2",
                              "selectExpr": {
                                "operand": {
                                  "id": "1",
                                  "identExpr": {
                                    "name": "R"
                                  }
                                },
                                "field": "attr"
                              }
                            },
                            "field": "owner"
                          }
                        },
                        {
                          "id": "6",
                          "selectExpr": {
                            "operand": {
                              "id": "5",
                              "identExpr": {
                                "name": "P"
                              }
                            },
                            "field": "id"
                          }
                        }
                      ]
                    }
                  }
                }
              }
            }
          }
        },
        "rules": [
          {
            "name": "rule-001",
            "actions": {
              "create": {}
            },
            "derivedRoles": {
              "employee_that_owns_the_record": {}
            },
            "effect": "EFFECT_ALLOW"
          },
          {
            "name": "rule-002",
            "actions": {
              "view:*": {}
            },
            "derivedRoles": {
              "direct_manager": {},
              "employee_that_owns_the_record": {}
            },
            "effect": "EFFECT_ALLOW"
          },
          {
            "name": "rule-003",
            "actions": {
              "view:public": {}
            },
            "derivedRoles": {
              "any_employee": {}
            },
            "effect": "EFFECT_ALLOW"
          },
          {
            "name": "rule-004",
            "actions": {
              "approve": {}
            },
            "derivedRoles": {
              "direct_manager": {}
            },
            "condition": {
              "expr": {
                "original": "request.resource.attr.status == \"PENDING_APPROVAL\"",
                "checked": {
                  "referenceMap": {
                    "1": {
                      "name": "request"
                    },
                    "5": {
                      "overloadId": [
                        "equals"
                      ]
                    }
                  },
                  "typeMap": {
                    "1": {
                      "messageType": "cerbos.engine.v1.Request"
                    },
                    "2": {
                      "messageType": "cerbos.engine.v1.Request.Resource"
                    },
                    "3": {
                      "mapType": {
                        "keyType": {
                          "primitive": "STRING"
                        },
                        "valueType": {
                          "dyn": {}
                        }
                      }
                    },
                    "4": {
                      "dyn": {}
                    },
                    "5": {
                      "primitive": "BOOL"
                    },
                    "6": {
                      "primitive": "STRING"
                    }
                  },
                  "sourceInfo": {
                    "location": "<input>",
                    "lineOffsets": [
                      51
                    ],
                    "positions": {
                      "1": 0,
                      "2": 7,
                      "3": 16,
                      "4": 21,
                      "5": 29,
                      "6": 32
                    }
                  },
                  "expr": {
                    "id": "5",
                    "callExpr": {
                      "function": "_==_",
                      "args": [
                        {
                          "id": "4",
                          "selectExpr": {
                            "operand": {
                              "id": "3",
                              "selectExpr": {
                                "operand": {
                                  "id": "2",
                                  "selectExpr": {
                                    "operand": {
                                      "id": "1",
                                      "identExpr": {
                                        "name": "request"
                                      }
                                    },
                                    "field": "resource"
                                  }
                                },
                                "field": "attr"
                              }
                            },
                            "field": "status"
                          }
                        },
                        {
                          "id": "6",
                          "constExpr": {
                            "stringValue": "PENDING_APPROVAL"
                          }
                        }
                      ]
                    }
                  }
                }
              }
            },
            "effect": "EFFECT_ALLOW"
          }
        ],
        "schemas": {
          "principalSchema": {
            "ref": "cerbos:///complex_object.json"
          },
          "resourceSchema": {
            "ref": "cerbos:///complex_object.json"
          }
        }
      },
      {
        "rules": [
          {
            "name": "wildcard",
            "actions": {
              "*": {}
            },
            "roles": {
              "admin": {}
            },
            "effect": "EFFECT_ALLOW"
          }
        ],
        "schemas": {
          "principalSchema": {
            "ref": "cerbos:///complex_object.json"
          },
          "resourceSchema": {
            "ref": "cerbos:///complex_object.json"
          }
        }
      }
    ],
    "schemas": {
      "principalSchema": {
        "ref": "cerbos:///complex_object.json"
      },
      "resourceSchema": {
        "ref": "cerbos:///complex_object.json"
      }
    }
  },
  "compilerVersion": 1
}
